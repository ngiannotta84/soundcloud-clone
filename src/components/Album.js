import React from "react";
import PropTypes from "prop-types";
import "../styles/album.css";
import { Link } from "react-router-dom";
import { music } from "../media/icons";

const Album = ({
  artistName,
  albumName,
  albumArt,
  songs,
  handleSetPlaylist,
  albumUserId,
  userId,
  albumId,
}) => {
  const songData = (song) => {
    return {
      image: albumArt,
      artistName,
      albumName,
      songName: song.name,
      audio: song.url,
    };
  };

  const handleClickSong = (song, addNext = false) => {
    const data = songData(song);
    handleSetPlaylist(data, addNext);
  };

  const handleClickAlbum = (addNext = false) => {
    const { length } = songs;

    if (addNext) {
      for (let i = length - 1; i >= 0; i -= 1) {
        const data = songData(songs[i]);
        handleSetPlaylist(data, addNext);
      }
    } else {
      for (let i = 0; i < length; i += 1) {
        const data = songData(songs[i]);
        handleSetPlaylist(data, addNext);
      }
    }
  };

  return (
    <div className="album">
      <div className="album__info-container">
        <img
          src={albumArt || music}
          alt={`${albumName} cover art`}
          className="album__cover-art"
        />
        <div className="album__info">
          <h2 className="album__artist-name">
            <Link to={`/profile/${artistName}`}>{artistName}</Link>
          </h2>
          <h3 className="album__album-name">{albumName}</h3>
        </div>
        <div className="album__buttons">
          <button
            type="button"
            className="addSong-button addSong-large"
            onClick={() => handleClickAlbum()}
          >
            Add Album to queue
          </button>
          <button
            type="button"
            className="addSong-button addSong-large"
            onClick={() => handleClickAlbum(true)}
          >
            Play Album next
          </button>
          {userId !== null && userId === albumUserId && (
            <Link
              to={`/edit/${albumId}`}
              className="album__buttons__link addSong-large"
            >
              Edit
            </Link>
          )}
        </div>
      </div>
      <li className="album__songs">
        {songs.map((song) => {
          return (
            <ul key={song.id} className="album__song">
              <h4 className="album__song__song-name">{song.name}</h4>
              <div>
                <button
                  type="button"
                  className="addSong-button album__song__button"
                  onClick={() => handleClickSong(song)}
                >
                  Add to queue
                </button>
                <button
                  type="button"
                  className="addSong-button album__song__button"
                  onClick={() => handleClickSong(song, true)}
                >
                  Play next
                </button>
              </div>
            </ul>
          );
        })}
      </li>
    </div>
  );
};

Album.defaultProps = {
  albumUserId: null,
  userId: null,
  albumId: null,
};

Album.propTypes = {
  artistName: PropTypes.string.isRequired,
  albumName: PropTypes.string.isRequired,
  albumArt: PropTypes.string.isRequired,
  songs: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      position: PropTypes.string,
      url: PropTypes.string,
    })
  ).isRequired,
  handleSetPlaylist: PropTypes.func.isRequired,
  albumUserId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  userId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  albumId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

export default Album;
