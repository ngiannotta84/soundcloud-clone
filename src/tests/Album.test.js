import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import Album from "../components/Album";

describe("Album", () => {
  const validProps = {
    artistName: "artist",
    albumName: "album",
    albumArt: "album url",
    songs: [
      {
        id: "1",
        name: "song 1",
        position: "1",
        url: "song url 1",
      },
      {
        id: "2",
        name: "song 2",
        position: "0",
        url: "song url 2",
      },
    ],
    handleSetPlaylist: jest.fn(),
  };

  test("snapshot", () => {
    const { asFragment } = render(
      <Router>
        <Album {...validProps} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe("tests", () => {
    beforeEach(() => {
      render(
        <Router>
          <Album {...validProps} />
        </Router>
      );
    });

    test("renders correctly", () => {
      expect(screen.getByRole("img")).toHaveAttribute(
        "alt",
        `${validProps.albumName} cover art`
      );
      expect(screen.getByText(validProps.artistName)).toHaveAttribute(
        "href",
        `/profile/${validProps.artistName}`
      );
      expect(screen.getByText(validProps.albumName)).toBeInstanceOf(
        HTMLHeadingElement
      );
      expect(screen.getAllByText(/add to queue/i)).toHaveLength(
        validProps.songs.length
      );
      expect(screen.getAllByText(/play next/i)).toHaveLength(
        validProps.songs.length
      );
      validProps.songs.forEach((song) => {
        expect(screen.getByText(song.name)).toBeInstanceOf(HTMLHeadingElement);
      });
    });

    test("add to queue button", () => {
      const button = screen.getAllByText(/add to queue/i)[0];
      const data = {
        image: validProps.albumArt,
        artistName: validProps.artistName,
        albumName: validProps.albumName,
        songName: validProps.songs[0].name,
        audio: validProps.songs[0].url,
      };

      fireEvent.click(button);

      expect(validProps.handleSetPlaylist).toBeCalledWith(data, false);
    });

    test("play next button", () => {
      const button = screen.getAllByText(/play next/i)[0];
      const data = {
        image: validProps.albumArt,
        artistName: validProps.artistName,
        albumName: validProps.albumName,
        songName: validProps.songs[0].name,
        audio: validProps.songs[0].url,
      };

      fireEvent.click(button);

      expect(validProps.handleSetPlaylist).toBeCalledWith(data, true);
    });
  });
});
