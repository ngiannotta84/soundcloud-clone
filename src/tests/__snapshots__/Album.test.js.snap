// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album snapshot 1`] = `
<DocumentFragment>
  <div
    class="album"
  >
    <div
      class="album__info-container"
    >
      <img
        alt="album cover art"
        class="album__cover-art"
        src="album url"
      />
      <div>
        <h2
          class="album__artist-name"
        >
          artist
        </h2>
        <h3
          class="album__album-name"
        >
          album
        </h3>
      </div>
    </div>
    <div>
      <div
        class="album__song"
      >
        <h4
          class="album__song__song-name"
        >
          song 2
        </h4>
        <div>
          <button
            class="album__song__add-song"
            type="button"
          >
            Add to queue
          </button>
          <button
            class="album__song__add-song"
            type="button"
          >
            Play next
          </button>
        </div>
      </div>
      <div
        class="album__song"
      >
        <h4
          class="album__song__song-name"
        >
          song 1
        </h4>
        <div>
          <button
            class="album__song__add-song"
            type="button"
          >
            Add to queue
          </button>
          <button
            class="album__song__add-song"
            type="button"
          >
            Play next
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
